### ISSUE TITLE
Conjunct for deletion of atoms is not properly calculated

### ISSUE THREAD

**Issue by hanjoosten** [AmpersandTarski/Ampersand#746]
I used Ampersand-v3.9.1 [sentinel/pending_bug:0cf88cd8e*] to compile the following script:

```
CONTEXT Try54

PATTERN Twovalued
RULE twovalued : I[Position] = "valid" \/ "invalid"
VIOLATION (SRC I, TXT " is not a Position. Please pick 'valid' or 'invalid'.")
ENDPATTERN

POPULATION Position CONTAINS [ "valid", "invalid" ]

INTERFACE Twovalued : "_SESSION"
BOX [ probeer : V[SESSION*Position] ]
ENDCONTEXT
```

I expected that it is impossible to delete either one of the initial atoms. 
Instead, I can delete both atoms in the concept Position through the user interface.

To reproduce, follow these steps:

1. Compile the script and generate a prototype
2. Run the prototype
3. press the dustbin icon of the atom `valid` 
4. watch the atom `valid` being removed from the database, without any violation being triggered.

**Comment by hanjoosten** [comment]
# Diagnosis 

A conjunct is being derived from the rule twovalued. The expression of that conjunct can be observed in the sql dump file (switch to use: --sqldump). In this case the expression is 

```
-I[Position] \/ ("valid"[Position] \/ "invalid"[Position])
```

**Comment by hanjoosten** [comment]
The sql derived for that expression is:

```sql
select
  distinct
         t1.src as src, t1.tgt as tgt
  from (/* case: EDif (l,r) I[Position] - "valid"[Position] ([Position*Position]) */
        /* Optimized case for: <expr1> intersect with the complement of <expr2>.
  where 
    <expr1> = I[Position] ([Position*Position])
    <expr2> = "valid"[Position] ([Position*Position])
     I[Position] /\ -"valid"[Position] ([Position*Position])
   */
        select t1.src as src, t1.tgt as tgt
        from (/* I[Position] */
              select "Position" as src, "Position" as tgt
              from "Position"
              where "Position" is not null)
             as t1
             left join (/* case: EMp1 val c */
                        select "Position" as src, "Position" as tgt
                        from "Position"
                        where "Position" = 'valid')
                       as t2
             on (t1.src = t2.src)
                and (t1.tgt = t2.tgt)
        where (t2.src is null)
              or (t2.tgt is null))
       as t1
       left join (/* case: EMp1 val c */
                  select "Position" as src, "Position" as tgt
                  from "Position"
                  where "Position" = 'invalid')
                 as t2
       on (t1.src = t2.src)
          and (t1.tgt = t2.tgt)
  where (t2.src is null)
        or (t2.tgt is null)
```

**Comment by hanjoosten** [comment]
@stefjoosten , I guess the normalizer is not delivering the right expression to the SQL-generator.

**Comment by hanjoosten** [comment]
Stef and I had a session, to diagnose this issue. It turns out, that the implementation of singleton expressions is wrong in the SQL generator. It should not look for the existance of the value in the database table. The SQL should return a single row containing the value in both the src as the tgt, regardless of presence in the database. The generated sql currently is:

```sql
/* case: EMp1 val c */
select "Position" as src, "Position" as tgt
from "Position"
where "Position" = 'valid'
```
It should look like:

```sql
/* case: EMp1 val c */
select 'valid' as src, 'valid' as tgt
from "ONE"
```

**Comment by hanjoosten** [comment]
It is **very** likely that this issue is connected with issue #342

**Comment by stefjoosten** [comment]
# Analysis
Consider rule
```
"valid"[Position] \/ "invalid"[Position] |- I[Position]
```
Its violations are computed as follows:
```
   -("valid"[Position] \/ "invalid"[Position] |- I[Position])
=
   -(-("valid"[Position] \/ "invalid"[Position]) \/ I[Position])
=
   ("valid"[Position] \/ "invalid"[Position]) /\ -I[Position]
```
Now suppose ``valid`` is being deleted from ``Position``, then we have:
```
   {(valid,valid), (invalid,invalid)} /\ -{(invalid,invalid)}
=
   {(valid,valid), (invalid,invalid)} /\ {}
=
   {}
```
This is not what we would expect.
We wanted the following violations:
```
   {(valid,valid), (invalid,invalid)} /\ {(invalid,invalid), (invalid,valid), (valid,invalid)}
=
   { (invalid,invalid) }
```
So apparently, the current algorithm is implemented correctly, but we want a different result.
This poses the question: why do we want a different result and how can we justify that in the theory?

**Comment by stefjoosten** [comment]
The compiler does a correct translation:
```
/*
Conjunct expression:
  (-"valid"[Position] /\ -"invalid"[Position]) \/ I[Position]
Rules for this conjunct:
  - twovalued: RULE "twovalued" : I[Position] = "valid"[Position] \/ "invalid"[Position]
     VIOLATION (SRC I[Position], TXT " is not a Position. Please pick 'valid' or 'invalid'.")
*/
select
  distinct
         someDummyNameBecauseMySQLNeedsOne.src as src,
         someDummyNameBecauseMySQLNeedsOne.tgt as tgt
  from (/* `intersect` does not work in MySQL, so this statement is generated: */
        select subIntersect0.src as src, subIntersect0.tgt as tgt
        from (/* case: EUni (l,r)"valid"[Position] \/ "invalid"[Position] ([Position*Position]) */
              /* case: EMp1 val c */
              select "valid" as src, "valid" as tgt
              from "ONE"
              union
              /* case: EMp1 val c */
              select "invalid" as src, "invalid" as tgt
              from "ONE"
             as subIntersect0,
             (/* case: ECpl (EDcI Position) */
              select "concept0"."Position" as src, "concept1"."Position" as tgt
              from "Position" as "concept0",
                   "Position" as "concept1"
              where "concept0"."Position" <> "concept1"."Position")
             as subIntersect1
        where (subIntersect1.src = subIntersect0.src)
              and (subIntersect1.tgt = subIntersect0.tgt))
       as someDummyNameBecauseMySQLNeedsOne
```

**Comment by hanjoosten** [comment]
fyi: I just started a new branch for this issue. The new generation of Mp1 is in place.

**Comment by hanjoosten** [comment]
I have done some more investigation. I come to the conclusion that there is a bug in MySQL. I [asked a question about this at the forum](https://forums.mysql.com/read.php?132,663749).

**Comment by RieksJ** [comment]
@stefjoosten: I am confused by what you wrote a few notes before. 

**QUOTE**
Now suppose `valid` is being deleted from Position, then we have:
~~~
   {(valid,valid), (invalid,invalid)} /\ -{(invalid,invalid)}
=
   {(valid,valid), (invalid,invalid)} /\ {}
=
   {}
~~~
This is not what we would expect.
We wanted the following violations:
~~~
   {(valid,valid), (invalid,invalid)} /\ {(invalid,invalid), (invalid,valid), (valid,invalid)}
=
   { (invalid,invalid) }
~~~
**ENDQUOTE**

I can only make sense of this derivation if I assume that
~~~
-{(invalid,invalid)} = {(invalid,invalid), (invalid,valid), (valid,invalid)}
~~~
However, I thought that
~~~
   -{(invalid,invalid)}
= 
   V[Position] - {(invalid,invalid)}
=
   {(valid,valid), (invalid,valid), (valid,invalid)}
~~~
which is not the same. Does this mean that your conclusion is incorrect?

**Comment by Michiel-s** [comment]
@hanjoosten, when I execute the exact queries you've posted on the forum, I just get one single row.

![image](https://user-images.githubusercontent.com/7961582/35197774-135c4884-fee5-11e7-82c6-bffd7d6015e3.png)

My setup:
`mysqld  Ver 10.1.21-MariaDB for Win32 on AMD64 (mariadb.org binary distribution)`

**Comment by Michiel-s** [comment]
@hanjoosten you've mentioned to me that this issue is diagnosed right. Could you summarize and close this issue is applicable?

**Comment by hanjoosten** [comment]
The initial implementation is right. As the derivation of @stefjoosten shows, in the current theory of Ampersand, there is no violation if an atom is deleted. @stefjoosten and @sjcjoosten have been discussing this. The theory has to be extended, so we can work with infinte collections (like String, Number etc) too.

**Comment by hanjoosten** [comment]
For the record: I will not change the behaviour of generation of MP1, since that is correct after all. That also means that ampersand does not generate queries that trigger the discoverd bug in MySQL.
